File: src/main/java/komu/demodel/domain/ModuleType.java
/*
 * Copyright (C) 2010 Juha Komulainen. All rights reserved.
 */
package komu.demodel.domain;

public enum ModuleType {
    PACKAGE {
        @Override
        public Module createModule(String name, PackageModule parent) {
            return new PackageModule(name, parent);
        }
    },
    
    TYPE {
        @Override
        public Module createModule(String name, PackageModule parent) {
            return new ClassModule(name, parent);
        }
    };

    public abstract Module createModule(String name, PackageModule parent);
}

File: src/main/scala/komu/demodel/domain/ModuleType.scala
/*
 * Copyright (C) 2010 Juha Komulainen. All rights reserved.
 */
package komu.demodel.domain

sealed abstract class ModuleType {
  def createModule(name: String, parent: PackageModule): Module
}

object ModuleType {
  case object PACKAGE extends ModuleType {
    def createModule(name: String, parent: PackageModule) = new PackageModule(name, Some(parent))
  }
  
  case object TYPE extends ModuleType {
    def createModule(name: String, parent: PackageModule) = new ClassModule(name, parent)
  }
}
